src/lights/AmbientLight.js

```diff
@@ -6,7 +6,7 @@ function AmbientLight( color, intensity ) {
 
 	this.type = 'AmbientLight';
 
-	this.castShadow = undefined;
+	this.castShadow = false;
```

Comment:
`castShadow` was set to `undefined` on purpose. It's not intended that this value is ever changed. 

This is also true for `HemisphereLight`.

Reply:
then typing is wrong.

Reply:
im curious that why it is overrided by `undefined`

Reply:
Because `AmbientLight` does not cast shadows. 

BTW: You can also change `HemisphereLight.d.ts`.

Reply:
https://github.com/mrdoob/three.js/pull/20340#discussion_r487577626

I made this decision because the base class has already initialized the prop by `false`, 
https://github.com/mrdoob/three.js/blob/fad6ce245763f37cd846f86e94b3ae3adfec227e/src/core/Object3D.js#L100 
which means the prop cannot be treated as optional, derived classes should override this prop (if needed) by either `true` or `false`, but not `undefined`, does this make sense?

Reply:
> which means the prop cannot be treated as optional, derived classes should override this prop (if needed) by either `true` or `false`, but not `undefined`, does this make sense? 

Is that a limitation of TypeScript?

Reply:
nope. TS can model concept of "absence" via `undefined | ...` or just `x?: ....`.

Reply:
I don't understand what's the problem then.

Reply:
`.castShadow` is currently modeled as a `boolean` in TS, which isn't correct due to this
https://github.com/mrdoob/three.js/blob/2ca2deb645e4a34747b34aa73ab3b34df47d72d5/src/lights/AmbientLight.js#L9

Reply:
And if we update the `.castShadow` type in AmbientLight.d.ts, it gets into this trouble https://github.com/mrdoob/three.js/pull/20340#discussion_r487747710

Reply:
Illustrate the issue if i correct the type to `.castShadow: undefined | boolean` for Object3D:

*Before* (`.castShadow: boolean`)

User code written in **.ts** 
```ts
if (mesh.castShadow === false) { /* assert: mesh does not cast shadow */ }
else { /* assert: mesh cast shadow */ }
```

*After* (`.castShadow: undefined | boolean`)

Same user code: 
```ts
if (mesh.castShadow === false) { /* assert: mesh does not cast shadow */ }
else { /* assert: mesh cast shadow <--- LOGIC ERROR : mesh.castShadow can be `undefined` */}
```

It passes typecheck but will fail at runtime.

Reply:
Can the property be deleted from the derived classes for which it does not apply?

(In suggesting this, I am assuming you do not want to change `Object3D`.)

Reply:
TBH, the concept of inheritance makes not really sense if subclass delete properties and methods as they like. The idea of deriving of something is that you can really that a certain interface exists.

It is possible in JS to set properties and methods to `undefined` but that does not mean it is advisable for all use cases. In this particular one, I vote for not setting `castShadow` to `undefined` but to `false`.

I mean you could also set `frustumCulled` to `undefined` for all lights because it is only evaluate for meshes, lines points and sprites.

Reply:
> 
> 
> Can the property be deleted from the derived classes for which it does not apply?
> 
> (In suggesting this, I am assuming you do not want to change `Object3D`.)

I just checked the AmbientLight heritages,
```
Object3d (.castShadow = false)
   > Light (no .castShadow accessor at all) 
       > AmibentLight (.castShadow = false (again))
```
It really makes sense to remove `this.castShadow = false` from AmbientLight.js.

Reply:
> It really makes sense to remove this.castShadow = false from AmbientLight.js.

I suggested this already here: https://github.com/mrdoob/three.js/pull/20340#discussion_r487913447

Reply:
@WestLangley  said:

>Can the property be deleted from the derived classes for which it does not apply?

>(In suggesting this, I am assuming you do not want to change Object3D.)

Since the property does not belong on `Object3D` in the first place, if you are going to fix it, I suggest you fix it correctly - instead of debating `false` vs. `undefined`.

Reply:
> Since the property does not belong on Object3D in the first place, if you are going to fix it, I suggest you fix it correctly 

In this case, `frustumCulled`, `castShadow`, `receiveShadow` and `renderOrder` needs to be refactored, too.

Reply:
I think this is my fault.

If I remember correctly, I set `this.castShadow = undefined;` so in the editor wouldn't create a checkbox for it. Happy to delete this line and change the editor so it checks the light type.

Reply:
Sounds good! 👍